### PR TITLE
Use acp 118 handler

### PR DESCRIPTION
**Pull request by ceyonur**
This pull request introduces several changes to the `plugin/evm/vm.go` and related files to improve caching, message handling, and testing in the warp backend. The idea is to merge caches in favor of Cached ACP-118 handler. It also refactors warp backend and implements acp118 verifier interface.
### Dependency Updates
* Updated the version of `github.com/ava-labs/avalanchego` in `go.mod` to the latest release. (`[go.modL8-R8](diffhunk://#diff-33ef32bf6c23acb95f5902d7097b7a1d5128ca061167ec0716715b0b9eeaa5f6L8-R8)`)
  * It depends on a version (master) that has acp118 handler caching changes. 

### Caching and Initialization Enhancements
* Added new imports for `cache` and `metercacher` modules in `plugin/evm/vm.go` to support enhanced caching mechanisms. (`[plugin/evm/vm.goR19-R22](diffhunk://#diff-3be5ea9687f8af021d9baaa6584d174a7975b181b604e1a759e4ad2e40136912R19-R22)`)
* Replaced the direct use of `warpSignatureCacheSize` with a `meteredCache` for caching warp signatures. This change improves cache management and monitoring. (`[plugin/evm/vm.goR491-R532](diffhunk://#diff-3be5ea9687f8af021d9baaa6584d174a7975b181b604e1a759e4ad2e40136912R491-R532)`)
* Refactored the `Initialize` method to include the creation and setup of `meteredCache` and added logic to clear the warp database if configured. (`[plugin/evm/vm.goR491-R532](diffhunk://#diff-3be5ea9687f8af021d9baaa6584d174a7975b181b604e1a759e4ad2e40136912R491-R532)`)
* Removed the `initializeHandlers` method and integrated its logic directly into the `Initialize` method, simplifying the initialization process. (`[plugin/evm/vm.goL619-L632](diffhunk://#diff-3be5ea9687f8af021d9baaa6584d174a7975b181b604e1a759e4ad2e40136912L619-L632)`)

### Testing Improvements
* Added new imports in `plugin/evm/vm_warp_test.go` to support enhanced testing capabilities. (`[[1]](diffhunk://#diff-816005c76cbcf9a7b2e47bb3b79f791507ae3b21594cfbbad80318ce065d2e2dR17)`, `[[2]](diffhunk://#diff-816005c76cbcf9a7b2e47bb3b79f791507ae3b21594cfbbad80318ce065d2e2dR39)`)
* Modified tests to use a fixed-size byte array for known signatures to ensure consistency in test results. (`[[1]](diffhunk://#diff-816005c76cbcf9a7b2e47bb3b79f791507ae3b21594cfbbad80318ce065d2e2dR756-R765)`, `[[2]](diffhunk://#diff-816005c76cbcf9a7b2e47bb3b79f791507ae3b21594cfbbad80318ce065d2e2dR814-R823)`)
* Added a new test `TestClearWarpDB` to verify the functionality of clearing the warp database and ensuring messages are correctly pruned when configured. (`[plugin/evm/vm_warp_test.goR858-R914](diffhunk://#diff-816005c76cbcf9a7b2e47bb3b79f791507ae3b21594cfbbad80318ce065d2e2dR858-R914)`)

### Backend Refactoring
* Updated imports in `warp/backend.go` to include new dependencies for enhanced caching and message verification. (`[warp/backend.goR14-R26](diffhunk://#diff-be66b853aa52c434887db29553e039e5536b325b159de2d4df44d2e8fb345930R14-R26)`)
* Refactored the `Backend` interface and its implementation to use a unified `signatureCache` instead of separate caches for message and block signatures. This change simplifies the caching logic and improves maintainability. (`[[1]](diffhunk://#diff-be66b853aa52c434887db29553e039e5536b325b159de2d4df44d2e8fb345930L40-R49)`, `[[2]](diffhunk://#diff-be66b853aa52c434887db29553e039e5536b325b159de2d4df44d2e8fb345930L62-R62)`)
* Removed the `Clear` method from the backend and integrated its functionality into the VM initialization process. (`[warp/backend.goL117-L123](diffhunk://#diff-be66b853aa52c434887db29553e039e5536b325b159de2d4df44d2e8fb345930L117-L123)`)
* Added a new `signMessage` method to encapsulate the signing logic and ensure signatures are cached consistently. (`[warp/backend.goR199-R208](diffhunk://#diff-be66b853aa52c434887db29553e039e5536b325b159de2d4df44d2e8fb345930R199-R208)`)

### Test Cleanup
* Removed the `TestClearDB` test from `warp/backend_test.go` as its functionality is now covered by the new `TestClearWarpDB` test in `plugin/evm/vm_warp_test.go`. (`[warp/backend_test.goL38-R46](diffhunk://#diff-3a69b001767fa655e727b1d3763e0d128c2145db8b6dca70474c22aa08eb2a9fL38-R46)`)

These changes collectively enhance the performance, reliability, and test coverage of the warp backend in the `plugin/evm` module.

## How this was tested

Added new UTs and improved existing warp tests.

## How is this documented

Not needed.

### CODE REVIEW

File: warp/stats.go [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+func newVerifierStats() *verifierStats {
+	return &verifierStats{
+		messageParseFail:                     metrics.NewRegisteredCounter("message_parse_fail", nil),
+		addressedCallSignatureValidationFail: metrics.NewRegisteredCounter("addressed_call_signature_validation_fail", nil),
+		blockSignatureValidationFail:         metrics.NewRegisteredCounter("block_signature_validation_fail", nil),
+	}
+}
+
+func (h *verifierStats) IncAddressedCallSignatureValidationFail() {
```
Comment (ceyonur) [inline warp/stats.go:26]: not sure how to use this as there is no "invalid addressedcall msg". They're further parsed for a known payload; and if there is no known payload type, then they're invalid (but again it means they cannot be parsed).

File: warp/verifier_backend.go [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+		b.stats.IncMessageParseFail()
+		return &common.AppError{
+			Code:    ParseErrCode,
+			Message: fmt.Sprintf("unknown payload type: %T", p),
+		}
+	}
+	return nil
+}
+
+// verifyBlockMessage verifies the block message (payload.Hash)
```
Comment (darioush) [inline warp/verifier_backend.go:69]:
```suggestion
// verifyBlockMessage returns nil if blockHashPayload contains the ID
// of an accepted block indicating it should be signed by the VM.
```

File: warp/verifier_backend.go [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+		return &common.AppError{
+			Code:    VerifyErrCode,
+			Message: fmt.Sprintf("failed to get block %s: %s", blockID, err.Error()),
+		}
+	}
+
+	return nil
+}
+
+// verifyAddressedCall verifies the addressed call message
```
Comment (darioush) [inline warp/verifier_backend.go:84]:
```suggestion
// verifyAddressedCall returns nil if addressedCall is parseable to a known payload type and
// passes type specific validation, indicating it should be signed by the VM.
// Note currently there are no valid payload types so this call always returns common.AppError
// with ParseErrCode.
```

File: warp/backend.go [hunk 4]
```diff
@@ -131,100 +121,56 @@ func (b *backend) AddMessage(unsignedMessage *avalancheWarp.UnsignedMessage) err
...
+	return b.signMessage(unsignedMessage)
 }
 
-func (b *backend) ValidateMessage(unsignedMessage *avalancheWarp.UnsignedMessage) error {
-	// Known on-chain messages should be signed
-	if _, err := b.GetMessage(unsignedMessage.ID()); err == nil {
-		return nil
-	}
+func (b *backend) GetBlockSignature(blockID ids.ID) ([]byte, error) {
+	log.Debug("Getting block from backend", "blockID", blockID)
```
Comment (darioush) [inline warp/backend.go:149]:
```suggestion
	log.Debug("Getting block signature from backend", "blockID", blockID)
```
also seems this log would be more useful if we emitted some indicator of success/failure

File: warp/stats.go [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+// See the file LICENSE for licensing terms.
+
+package warp
+
+import (
+	"github.com/ava-labs/subnet-evm/metrics"
+)
+
+type verifierStats struct {
+	messageParseFail metrics.Counter
```
Comment (darioush) [inline warp/stats.go:11]: should we have some success metrics as well? (or are those available through the p2p handler?)

File: warp/backend.go [hunk 6]
```diff
@@ -248,3 +194,13 @@ func (b *backend) GetMessage(messageID ids.ID) (*avalancheWarp.UnsignedMessage,
 
 	return unsignedMessage, nil
 }
+
+func (b *backend) signMessage(unsignedMessage *avalancheWarp.UnsignedMessage) ([]byte, error) {
+	sig, err := b.warpSigner.Sign(unsignedMessage)
+	if err != nil {
+		return []byte{}, fmt.Errorf("failed to sign warp message: %w", err)
```
Comment (darioush) [inline warp/backend.go:201]: style nit: should we switch to returning `nil` in error cases instead of []byte{}?

File: warp/backend.go [hunk 7]
```diff
@@ -131,100 +121,56 @@ func (b *backend) AddMessage(unsignedMessage *avalancheWarp.UnsignedMessage) err
 		return fmt.Errorf("failed to put warp signature in db: %w", err)
 	}
 
-	var signature [bls.SignatureLen]byte
 	sig, err := b.warpSigner.Sign(unsignedMessage)
```
Comment (darioush) [inline warp/backend.go:124]: should we use b.signMessage here?

File: warp/backend.go [hunk 8]
```diff
@@ -131,100 +121,56 @@ func (b *backend) AddMessage(unsignedMessage *avalancheWarp.UnsignedMessage) err
...
 	if err != nil {
```
Comment (darioush) [inline warp/backend.go:125]: wonder if this function should check sourceChainID and networkID like we check for offchain messages

File: warp/handlers/signature_request.go [hunk 9]
```diff
@@ -51,13 +51,14 @@ func (s *SignatureRequestHandler) OnMessageSignatureRequest(ctx context.Context,
 		log.Debug("Unknown warp message requested", "messageID", signatureRequest.MessageID)
 		s.stats.IncMessageSignatureMiss()
 	} else {
-		signature, err = s.backend.GetMessageSignature(unsignedMessage)
+		sig, err := s.backend.GetMessageSignature(unsignedMessage)
 		if err != nil {
 			log.Debug("Unknown warp signature requested", "messageID", signatureRequest.MessageID)
 			s.stats.IncMessageSignatureMiss()
 			signature = [bls.SignatureLen]byte{}
```
Comment (darioush) [inline warp/handlers/signature_request.go:58]: can this line be removed?

File: warp/handlers/signature_request.go [hunk 10]
```diff
@@ -80,13 +81,15 @@ func (s *SignatureRequestHandler) OnBlockSignatureRequest(ctx context.Context, n
 		s.stats.UpdateBlockSignatureRequestTime(time.Since(startTime))
 	}()
 
-	signature, err := s.backend.GetBlockSignature(request.BlockID)
+	var signature [bls.SignatureLen]byte
+	sig, err := s.backend.GetBlockSignature(request.BlockID)
 	if err != nil {
 		log.Debug("Unknown warp signature requested", "blockID", request.BlockID)
 		s.stats.IncBlockSignatureMiss()
 		signature = [bls.SignatureLen]byte{}
```
Comment (darioush) [inline warp/handlers/signature_request.go:89]: can this line be removed?

File: plugin/evm/vm_warp_test.go [hunk 11]
```diff
@@ -849,3 +855,60 @@ func TestBlockSignatureRequestsToVM(t *testing.T) {
...
+		_, err = vm.warpBackend.GetMessageSignature(unsignedMsg)
+		require.NoError(t, err)
+		messages = append(messages, unsignedMsg)
+	}
+
+	require.NoError(t, vm.Shutdown(context.Background()))
+
+	// Restart VM with the same database default should not prune the warp db
+	vm = &VM{}
+	ctx, _, _, _, _ = setupGenesis(t, genesisJSONLatest)
```
Comment (darioush) [inline plugin/evm/vm_warp_test.go:885]: could we add a comment why this line is needed?

File: go.mod [hunk 12]
```diff
@@ -5,7 +5,7 @@ go 1.22.8
 require (
 	github.com/VictoriaMetrics/fastcache v1.12.1
 	github.com/antithesishq/antithesis-sdk-go v0.3.8
-	github.com/ava-labs/avalanchego v1.11.12-rc.2.0.20241003220930-43d5b435a644
+	github.com/ava-labs/avalanchego v1.11.13-0.20241018024240-22fb2f016633
```
Comment (darioush) [inline go.mod:8]: could you add a comment to the PR or in the description a link or summary of what has changed in avalanchego?

File: warp/stats.go [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+		addressedCallSignatureValidationFail: metrics.NewRegisteredCounter("addressed_call_signature_validation_fail", nil),
+		blockSignatureValidationFail:         metrics.NewRegisteredCounter("block_signature_validation_fail", nil),
+	}
+}
+
+func (h *verifierStats) IncAddressedCallSignatureValidationFail() {
+	h.addressedCallSignatureValidationFail.Inc(1)
+}
+
+func (h *verifierStats) IncBlockSignatureValidationFail() {
```
[the remark this answers is not in the document]
Comment (ceyonur) [reply warp/stats.go]: Not sure if we need to be that much specific/detailed as there is no other checks for block IDs.

File: warp/verifier_backend_test.go [hunk 14]
```diff
@@ -0,0 +1,257 @@
...
+				name += "_with_cache"
+			} else {
+				name += "_no_cache"
+			}
+			t.Run(name, func(t *testing.T) {
+				var sigCache cache.Cacher[ids.ID, []byte]
+				if withCache {
+					sigCache = &cache.LRU[ids.ID, []byte]{Size: 100}
+				} else {
+					sigCache = &cache.Empty[ids.ID, []byte]{}
```
[the remark this answers is not in the document]
Comment (ceyonur) [reply warp/verifier_backend_test.go:99]: I think it worth it since we deferred the cache management to avalanchego

**Comment by ceyonur** [comment]
Converted one this to "Draft" since we might release a new Subnet-EVM and this one still depends on avalanchego@master.

File: warp/verifier_backend.go [hunk 15]
```diff
@@ -0,0 +1,105 @@
...
+	VerifyErrCode
+)
+
+// Verify implements the acp118.Verifier interface
+func (b *backend) Verify(_ context.Context, unsignedMessage *avalancheWarp.UnsignedMessage, _ []byte) *common.AppError {
+	return b.verifyMessage(unsignedMessage)
+}
+
+// verifyMessage verifies the signature of the message
+// This is moved to a separate function to avoid having to use a context.Context
```
[the remark this answers is not in the document]
Comment (ceyonur) [reply warp/verifier_backend.go:28]: hmm yea I think we can just add ctx to every call place.

File: warp/verifier_backend.go [hunk 16]
```diff
@@ -0,0 +1,105 @@
...
+			return apperr
+		}
+	default:
+		b.stats.IncMessageParseFail()
+		return &common.AppError{
+			Code:    ParseErrCode,
+			Message: fmt.Sprintf("unknown payload type: %T", p),
+		}
+	}
+	return nil
```
[the remark this answers is not in the document]
Comment (ceyonur) [reply warp/verifier_backend.go:63]: done

File: warp/verifier_backend.go [hunk 17]
```diff
@@ -0,0 +1,105 @@
...
+
+	switch p := parsed.(type) {
+	default:
+		b.stats.IncMessageParseFail()
+		return &common.AppError{
+			Code:    ParseErrCode,
+			Message: fmt.Sprintf("unknown message type: %T", p),
+		}
+	}
+}
```
[the remark this answers is not in the document]
Comment (ceyonur) [reply warp/verifier_backend.go:105]: I will move it to uptime verification

**Review by darioush — APPROVED** [review]

File: warp/stats.go [hunk 18]
```diff
@@ -0,0 +1,36 @@
...
+	messageParseFail metrics.Counter
+	// AddressedCall metrics
+	addressedCallSignatureValidationFail metrics.Counter
+	// BlockRequest metrics
+	blockSignatureValidationFail metrics.Counter
+}
+
+func newVerifierStats() *verifierStats {
+	return &verifierStats{
+		messageParseFail:                     metrics.NewRegisteredCounter("message_parse_fail", nil),
```
[the remark this answers is not in the document]
Comment (ceyonur) [reply warp/stats.go]: I added it in https://github.com/ava-labs/subnet-evm/pull/1367

**Pull request merged**